### PR TITLE
Use `bech32` to calculate descriptor checksums

### PR DESCRIPTION
**Pull request by tcharding**
This is just the last patch, the rest are in #609 now.

The `checksum::poly_mod` function implements BCH codes to calculate a checksum (appended to descriptors). We recently released a version of BCH codes in `bech32`. We can implement the `bech32::Checksum` trait for BIP-380 and use the `primitives::checksum` module, removing the custom `poly_mod` function.

### CODE REVIEW

**Comment by apoelstra** [comment]
Up to 5018673bf5642850cd6aad3d25697de217eb105e is great and could be PR'd separately if you want to move more quickly.

**Comment by apoelstra** [comment]
As for 9ca3380f197f1bf93f92ff1c2729c7c70724b3c4:

* We should pull apart the `Engine` and replace the cls/clscount stuff with an iterator adaptor from `char`s to `Fe`s.
* `Osd` can be `DescriptorChecksum` or something. It doesn't have to be short; it shouldn't ever be used by users.
* I believe your `expect`s are all correct; you are construting a 3-digit base 3 number, which can be at most 222 (or 26 in decimal) which fits into a Fe32.
* Your "comment why we do this" TODO is because you're manually computing a checksum and manually futzing with the target checksum. Just use `Engine::input_target_residue` and then you can drop the code that inputs 0s and basically all the rest of this function. Also you can just replace the whole `checksum_chars` function with the `Checksummed` iterator adaptor from rust-bech32.

**Comment by tcharding** [comment]
Sick! Thanks man.

**Comment by tcharding** [comment]
I learned a new expression 'futzing with', not the first one I've learned from you @apoelstra, I love it.

File: src/descriptor/bare.rs
Comment (apoelstra) [inline src/descriptor/bare.rs:168]: In 22ee854c61b13d15fb7d0b11daf2a87604f45ba7:

We definitely want to keep the `wrapped_f` stuff. This new code is both longer and now allocates.

File: src/descriptor/checksum.rs
Comment (apoelstra) [inline src/descriptor/checksum.rs:41]: In 22ee854c61b13d15fb7d0b11daf2a87604f45ba7:

Rather than taking a string this should take a `D: Display` and a `fmt::Formatter`.

File: src/descriptor/checksum.rs
Comment (apoelstra) [inline src/descriptor/checksum.rs:176]: In 22ee854c61b13d15fb7d0b11daf2a87604f45ba7:

It would be faster to feed the checksum into the checksum iterator and compare the residue against the target. Then in the happy case we only need to iterate once.

In the unhappy case, we should probably return a better error variant that has the expected and received checksum in it. It's fine if this is slow and allocating.

Comment (tcharding) [reply src/descriptor/checksum.rs:176]: Sure thing.

File: src/descriptor/bare.rs
Comment (tcharding) [reply src/descriptor/bare.rs:168]: Yeah I balked at this before I pushed it. Will add it back in.

**Comment by tcharding** [comment]
Alright, I think I got a little trigger happy before to use my new found iterator skills from the recent bech32 work. Instead this PR now just does the minimum required to remove the `poly_mod` function and use `bech32::primitives::checksum` instead.

**Comment by tcharding** [comment]
I was futzing with the whole `descriptor` module before ;)

**Comment by apoelstra** [comment]
Ok, ACK this. It's a good start.

**Comment by tcharding** [comment]
Rebased to pick up changes in #609 and use newly release `bech32 v0.10.0-beta`, no other changes.

**Review by apoelstra — DISMISSED** [review]
ACK c5c08fe43a914b0ae138f598f0c36725d161e7a4

**Comment by apoelstra** [comment]
@sanket1729 I'll let you take a look at this before merging. It basically gets rid of some of the magic constants and uses rust-bech32 facilities to do the checksum. But this doesn't represent a "real" use of the new rust-bech32 API, which would pretty-much reduce the checksum.rs module to just encoding/decoding strings as u5s and then handing off to rust-bech32 APIs.

**Comment by tcharding** [comment]
I was going to attack it in steps:
- Move all the `Formatter` stuff in `descriptor/*.rs` to use a macro `write_desciptor!`.
- Then in the macro work out how to convert the input string to a stream of `Fe32`s like you suggest.

The reason to do two steps is that I don't think we can have a general "input multiple strings" to the checksum engine because cls/clscount is state that has to be kept between successive input strings. But i noticed that we never do more than input a single descriptor.

**Comment by apoelstra** [comment]
I don't like the idea of moving so much logic into a macro.

If we need to retain state then we can keep the existing `Engine` structure with `cls`/`clscount`. If we don't, we should change this to be an iterator adaptor.

But either way, once we have a stream of Fe32s, we should be able to just put them into a rust-bech32 `checksum::Engine` and everything should be easy.

**Comment by tcharding** [comment]
> I don't like the idea of moving so much logic into a macro.

The logic is just the more or less duplicate lines 
```rust
        use fmt::Write;
        let mut wrapped_f = checksum::Formatter::new(f);
        write!(wrapped_f, "{}", self.ms)?;
        wrapped_f.write_checksum_if_not_alt()
```
Other descriptors have a match statement when calling `write` but they are essentially the same. Its totally just an abstraction thing, as it is it appears sane to call `write` multiple times, and the current implementation supports this. But its an unnecessary generalisation I believe because we only ever call `write` once, to write a single descriptor. 

> If we need to retain state then we can keep the existing `Engine` structure with `cls`/`clscount`. If we don't, we should change this to be an iterator adaptor.
>
> But either way, once we have a stream of Fe32s, we should be able to just put them into a rust-bech32 `checksum::Engine` and everything should be easy.

Another complication is we cannot use bech32 iterators to encode because of the `#` symbol. That is what led me to create the  `Checksummed` type. Anyways, I'll have another play with it.

**Comment by apoelstra** [comment]
>The logic is just the more or less duplicate lines

Oh, yeah, this seems fine to me.

>Another complication is we cannot use bech32 iterators to encode because of the # symbol.

Yeah, that sucks. But I think your solution is good.

**Comment by tcharding** [comment]
I had another look at this, I'm not sure we can do much better even using some sort of iterator. I'm pretty sure that our usage of `fmt::Write::write_str` is doing an allocation under the hood. Changing our usage is then clearly an optimization problem not a "use bech32" problem.

**Comment by tcharding** [comment]
I think this should be considered for merge as is, I wrote up #610 to discuss potential improvements.

**Comment by apoelstra** [comment]
>I'm not sure we can do much better even using some sort of iterator

Yeah, you may be right.

**Review by apoelstra — APPROVED** [review]
ACK c44ce81d94e80cd1c814146d9b6e91c7def3ff9a

**Review by sanket1729 — APPROVED** [review]
utACK c44ce81d94e80cd1c814146d9b6e91c7def3ff9a. Did not review the use of bech32 carefully. Any breakage there should have caused existing test cases to fail.

**Comment by sanket1729** [comment]
We can also safely merge this despite the red crosses as CI is fixed in master.

**Comment by tcharding** [comment]
I rebased it for good measure, no changes.

**Review by apoelstra — APPROVED** [review]
ACK fe1a6be3bba92ed7c27d5bc8e72dc590f2e4dc88

**Pull request merged**
